### PR TITLE
Locate tag selection & rewrite function inside XmlEngine

### PR DESCRIPTION
**Pull request by agb94**

### CODE REVIEW

**Comment by agb94** [comment]
added `postproc_tree` method.. or could you suggest any better name?... 😬

**Comment by bloa** [comment]
This is much better! :+1: 
I would simply call it `process_tree` rather than `postproc_tree`.
Also I wonder if it would not be worth putting empty `process_tree` and `process_modification_points` directly in AbstractProgram.

Why is `rotate_newlines` still in `get_contents` and not in the new method?

**Comment by agb94** [comment]
> Also I wonder if it would not be worth putting empty process_tree and process_modification_points directly in AbstractProgram.
I think `process_tree` is a kind of a special feature of XmlEngine because LineEngine or LineProgram does not need to call `process_tree` method after loading the contents.

> Why is rotate_newlines still in get_contents and not in the new method?
I moved it to the process_tree method!

**Comment by agb94** [comment]
And could you explain what the role of `rotate_newlines` method is? I don't quite understand the code 😪

**Comment by bloa** [comment]
(do not accept pull request if you do not understand what the code does! I hope you checked XML edits...)

TLDR: you can remove it, it is a nice but not indispensable method.

Lets say we have the following source XML fragment: (kinda similar to what srcML produces)
```
    <stmt>a = 1;</stmt>
    <stmt>b = 2;</stmt>
    <stmt>c = 3;</stmt>
    
```
Then deletion or insertion will happily make a mess of newlines between statements.
For example: (insertion of stmt2 before stmt1, deletion of stmt2)
```
    <stmt>b = 2;</stmt><stmt>a = 1;</stmt>
    
    <stmt>c = 3;</stmt>
```
```
    b = 2;a = 1;
    
    c = 3;
```
This might be okay for say C or Java, but with for example Python (with no semicolon) everything will break.
This will however not happen if newlines are included inside the tags:
```
    <stmt>a = 1;
    </stmt><stmt>b = 2;
    </stmt><stmt>c = 3;
</stmt>
```
Produces with the same patch:
```
    <stmt>b = 2;
    </stmt><stmt>a = 1;
    </stmt><stmt>c = 3;
</stmt>
```
```
    b = 2;
    a = 1;
    c = 3;
```

`rotate_newlines` is a small helper method that do exactly this.
This is mainly for cosmetics reasons, definitely not perfect, and have not been really proven to always work. Btw the regexp should be "(^\n\s*)" and not "(\n\s*)" (probably with a raw string also?).
I think it would be better if indentation was better handled (included before, not after).

Note that for the exact same reason, in https://github.com/coinse/pyggi/blob/version2/sample/Triangle_fast_xml/Triangle.java.xml I put
```
<line>    public enum TriangleType {
</line><line>        INVALID, SCALENE, EQUALATERAL, ISOCELES
</line><line>    }
</line>
```
Rather than
```
<line>    public enum TriangleType {</line>
<line>        INVALID, SCALENE, EQUALATERAL, ISOCELES</line>
<line>    }</line>
```

**Comment by agb94** [comment]
@bloa Now I got it. Thanks for the detailed explanation.
I'd looked into the code snippet and checked whether there are no side effects but did not understand what's the precise intention of the method. I should've asked about it before merging the previous pull request, but I suspended it bc we'll have the final pull request from version2 to master.. I will carefully look into every changes again before releasing v2!
And I've also tried to handle the indentation problem in the `do_insertion` method by making the indentation level of the inserted node identical to one of its siblings w/o `rotate_newlines` method, but that doesn't seem like a perfect solution either. :( will try to find better solutions
+ please check the new commit and tell me if any further changes are needed

**Comment by agb94** [comment]
@bloa please check the script changes :)

**Comment by bloa** [comment]
It looks that classes in examples are finally at top level (compared to inside the `if __name__ == "__main__":`).
Do you think it could be possible to put them in a `helper.py` to avoid code duplication? For example code relative to computing fitness or the local search.

BTW I have no idea how the `operators` variable works, is it used as a global variable?
If yes, then it would be probably much better to simply set them through a property of the `MyLocalSearch` class.

**Comment by agb94** [comment]
> Do you think it could be possible to put them in a helper.py to avoid code duplication? For example code relative to computing fitness or the local search.

I'll work on it after filling out the documentation ..!! :)
